contracts/implementation/KeeperRewards.sol

```diff
@@ -0,0 +1,165 @@
+         * +---------------------------+------+---+---+~~~~~
+         *
+         * Under normal circumstances, we don't expect there to be any time
+         * premium at all. The time premium exists in order to *further*
+         * incentivise upkeep in the event of lateness.
+         *
+         * The base tip exists to act as pure profit for a keeper.
+         *
+         * Of course, the gas component acts as compensation for performing
+         * on-chain computation.
+         *
+         */
+
+        // keeper gas cost in wei. WAD formatted
+        uint256 _keeperGas = keeperGas(_gasPrice, _gasSpent, _settlementTokenPrice);
+
+        // tip percent
+        uint256 _tipPercent = keeperTip(_savedPreviousUpdatedTimestamp, _poolInterval);
+
+        // amount of settlement tokens to give to the keeper
+        // _keeperGas + _keeperGas * percentTip
+        uint256 wadRewardValue = _keeperGas + ((_keeperGas * _tipPercent) / 100);
+
+        return wadRewardValue;
+    }
+
+    /**
+     * @notice Compensation a keeper will receive for their gas expenditure
+     * @param _gasPrice Price of a single gas unit (in ETH (wei))
+     * @param _gasSpent Number of gas units spent
+     * @return Keeper's gas compensation
+     * @dev Adds a constant to `_gasSpent` when calculating actual gas usage
+     */
+    function keeperGas(
+        uint256 _gasPrice,
+        uint256 _gasSpent,
+        uint256 _settlementTokenPrice
+    ) public pure returns (uint256) {
+        if (_settlementTokenPrice <= 0) {
+            return 0;
+        } else {
+            /* gas spent plus our fixed gas overhead */
+            uint256 gasUsed = _gasSpent + FIXED_GAS_OVERHEAD;
+
+            /* safe due to explicit bounds check for settlementTokenPrice above */
+            /* (wei * Settlement / ETH) / fixed point (10^18) = amount in settlement */
+            bytes16 weiSpent = ABDKMathQuad.fromUInt(_gasPrice * gasUsed);
+            bytes16 settlementTokenPrice = ABDKMathQuad.fromUInt(uint256(_settlementTokenPrice));
+            return ABDKMathQuad.toUInt(ABDKMathQuad.div(ABDKMathQuad.mul(weiSpent, settlementTokenPrice), FIXED_POINT));
+        }
+    }
+
+    /**
+     * @notice Tip a keeper will receive for successfully updating the specified pool
+     * @param _savedPreviousUpdatedTimestamp Last timestamp when the pool's price execution happened
+     * @param _poolInterval Pool interval of the given pool
+     * @return Percent of the `keeperGas` cost to add to payment, as a percent
+     */
+    function keeperTip(uint256 _savedPreviousUpdatedTimestamp, uint256 _poolInterval) public view returns (uint256) {
+        /* the number of blocks that have elapsed since the given pool's updateInterval passed */
+        uint256 elapsedBlocksNumerator = (block.timestamp - (_savedPreviousUpdatedTimestamp + _poolInterval));
+
+        uint256 keeperTipAmount = BASE_TIP + (TIP_DELTA_PER_BLOCK * elapsedBlocksNumerator) / BLOCK_TIME;
```

Comment:
interested as to why we are using L1 block times here (or is it just coincidence?). 

we could make the number less "magic" by doing something like:

```
uint256 secondsOverdue = (block.timestamp - (_savedPreviousUpdatedTimestamp + _poolInterval));
uint256 percentOverdue = (secondsOverdue * 100) / _poolInterval;

uint256 keeperTipAmount = BASE_TIP + (TIP_DELTA_PER_PERCENT_OVERDUE * percentOverdue);
```

Reply:
Good point. out of scope

Reply:
yep agree, good thing we can change this now